lib/logstash/filters/memcached.rb

```diff
@@ -141,20 +164,21 @@ def do_set(event)
     return true
   end
 
-  def establish_connection
-    require 'dalli'
-
-    hosts = validate_connection_hosts
-    options = validate_connection_options
+  def new_connection(hosts, options)
     logger.debug('connecting to memcached', context(hosts: hosts, options: options))
-    Dalli::Client.new(@hosts, options).tap do |client|
-      begin
-        client.alive!
-      rescue Dalli::RingError
-        logger.error("failed to connect", context(hosts: hosts, options: options))
-        fail("cannot connect to memcached")
-      end
+    Dalli::Client.new(hosts, options).tap { |client| client.alive! }
+  end
+
+  def reconnect(hosts, options)
+    begin
+      @cache = new_connection(hosts, options)
```

Comment:
this method should only be executed in mutual exclusion, since it mutates shared state.

If we fail to do so, it _may_ work, but it will also mean each disconnection will be coupled with a flood of independent reconnects coming from each of the workers.

Reply:
as per my main thread comment, I totally overlooked thread safety. I will correct that.